File: app/src/main/java/com/majeur/psclient/io/GlideHelper.java
package com.majeur.psclient.io;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.view.ViewGroup;
import android.view.ViewPropertyAnimator;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.bumptech.glide.request.target.ViewTarget;
import com.majeur.psclient.R;
import com.majeur.psclient.model.BasePokemon;
import com.majeur.psclient.model.BattlingPokemon;
import com.majeur.psclient.model.Player;
import com.majeur.psclient.util.Utils;
import com.majeur.psclient.util.html.Html;

import java.util.concurrent.ExecutionException;

import static com.majeur.psclient.model.Player.FOE;
import static com.majeur.psclient.util.Utils.array;

public class GlideHelper {

    static {
        ViewTarget.setTagId(R.id.glide_tag);
    }

    private RequestManager mRequestManager;

    public GlideHelper(Context context) {
        mRequestManager = Glide.with(context);
    }

    private final float MAGIC_SCALE = 0.0027777777777778f;

    @SuppressWarnings("CheckResult")
    public void loadSprite(final BattlingPokemon pokemon, final ImageView imageView, final int fieldWidth) {
        RequestBuilder<Drawable> request = mRequestManager.load(spriteUri(pokemon.spriteId, pokemon.foe, pokemon.shiny));
        RequestOptions options = new RequestOptions().override(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL);
        request.apply(options);
        request.error(
                mRequestManager.load(spriteUriAlt(pokemon.spriteId, pokemon.foe, pokemon.shiny))
                .error(
                        mRequestManager.load(spriteUriAltAlt(pokemon.spriteId, pokemon.foe, pokemon.shiny))
                        .apply(options.error(R.drawable.missingno))
                )
        );
        request.into(new AnimatedImageViewTarget(imageView) {
            @Override
            protected void onInitInAnimation(ViewPropertyAnimator viewPropertyAnimator) {
                viewPropertyAnimator
                        .setDuration(250)
                        .setInterpolator(new DecelerateInterpolator())
                        .scaleX(0f)
                        .scaleY(0f)
                        .alpha(0f);
            }

            @Override
            protected void onInitOutAnimation(ViewPropertyAnimator viewPropertyAnimator) {
                viewPropertyAnimator
                        .setDuration(250)
                        .setInterpolator(new AccelerateInterpolator())
                        .scaleX(1f)
                        .scaleY(1f)
                        .alpha(1f);
            }

            @Override
            protected void setResource(Drawable resource) {
                int scale = Math.round(fieldWidth * MAGIC_SCALE);
                if (!pokemon.foe) scale = Math.round(scale * 1.5f);

                ViewGroup.LayoutParams layoutParams = imageView.getLayoutParams();
                layoutParams.width = resource.getIntrinsicWidth() * scale;
                layoutParams.height = resource.getIntrinsicHeight() * scale;

                imageView.setImageDrawable(resource);
            }
        });
    }

    public void loadPreviewSprite(Player player, BasePokemon pokemon, ImageView imageView) {
        RequestBuilder<Drawable> request = mRequestManager.load(spriteUri(pokemon.spriteId, player == FOE, false));
        request.apply(new RequestOptions().error(R.drawable.missingno));
        ViewGroup.LayoutParams layoutParams = imageView.getLayoutParams();
        layoutParams.width = layoutParams.height = ViewGroup.LayoutParams.WRAP_CONTENT;
        request.into(imageView);
    }

    public void loadDexSprite(BasePokemon pokemon, boolean shiny, ImageView imageView) {
        RequestBuilder<Drawable> request = mRequestManager.load(dexSpriteUri(pokemon.spriteId, shiny));
        request.apply(new RequestOptions().error(R.drawable.placeholder_pokeball));
        request.into(imageView);
    }

    public void loadAvatar(String avatar, ImageView imageView) {
        RequestBuilder<Drawable> request = mRequestManager.load(avatarSpriteUri(avatar));
        request.into(imageView);
    }

    private StringBuilder baseUri() {
        return new StringBuilder()
                .append("https://play.pokemonshowdown.com/sprites/");
    }

    private String spriteUri(String spriteId, boolean foe, boolean shiny) {
        return baseUri()
                .append(foe ? "ani" : "ani-back")
                .append(shiny ? "-shiny/" : "/")
                .append(spriteId)
                .append(".gif")
                .toString();
    }

    private String spriteUriAlt(String spriteId, boolean foe, boolean shiny) {
        return baseUri()
                .append(foe ? "gen5ani" : "gen5ani-back")
                .append(shiny ? "-shiny/" : "/")
                .append(spriteId)
                .append(".png")
                .toString();
    }

    private String spriteUriAltAlt(String spriteId, boolean foe, boolean shiny) {
        return baseUri()
                .append(foe ? "gen5" : "gen5-back")
                .append(shiny ? "-shiny/" : "/")
                .append(spriteId)
                .append(".png")
                .toString();
    }

    private String dexSpriteUri(String spriteId, boolean shiny) {
        return baseUri()
                .append("dex")
                .append(shiny ? "-shiny/" : "/")
                .append(spriteId)
                .append(".png")
                .toString();
    }

    private String typeSpriteUri(String type) {
        return baseUri()
                .append("types/")
                .append(Utils.firstCharUpperCase(type))
                .append(".png")
                .toString();
    }

    private String categorySpriteUri(String category) {
        return baseUri()
                .append("categories/")
                .append(Utils.firstCharUpperCase(category))
                .append(".png")
                .toString();
    }

    private String avatarSpriteUri(String avatar) {
        return baseUri()
                .append("trainers/")
                .append(avatar)
                .append(".png")
                .toString();
    }

    public Html.ImageGetter getHtmlImageGetter(final DexIconLoader iconLoader, int maxWidth) {
        final int mw = maxWidth - Utils.dpToPx(2);
        return new Html.ImageGetter() {
            @Override
            public Drawable getDrawable(final String source, int reqw, int reqh) {
                try {
                    Drawable d = null;
                    if (source.startsWith("content://com.majeur.psclient/dex-icon/")) {
                        String species = source.substring(source.lastIndexOf('/') + 1, source.length());
                        Bitmap icon = iconLoader.load(array(species))[0];
                        if (icon != null) d = new BitmapDrawable(icon);
                    } else {
                        d = mRequestManager.asDrawable().load(source).submit().get();
                    }
                    if (d == null) return null;
                    float r = d.getIntrinsicWidth() / (float) d.getIntrinsicHeight();
                    int w, h;
                    if (reqw != 0 && reqh == 0) {
                        w = reqw;
                        h = (int) (w / r);
                    } else if (reqw == 0 && reqh != 0) {
                        h = reqh;
                        w = (int) (h * r);
                    } else {
                        w = reqw;
                        h = reqh;
                    }
                    float mr = w / (float) mw;
                    if (mr > 1) {
                        w = mw;
                        h /= mr;
                    }
                    d.setBounds(0, 0, w, h);
                    return d;
                } catch (ExecutionException | InterruptedException e) {
                    e.printStackTrace();
                    return null;
                }
            }
        };
    }
}

File: psclient/src/main/java/com/majeur/psclient/io/GlideHelper.kt
package com.majeur.psclient.io

import android.content.Context
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.view.ViewPropertyAnimator
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import android.widget.ImageView
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestBuilder
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.majeur.psclient.R
import com.majeur.psclient.model.battle.Player
import com.majeur.psclient.model.pokemon.BasePokemon
import com.majeur.psclient.model.pokemon.BattlingPokemon
import com.majeur.psclient.util.Utils
import com.majeur.psclient.util.glide.AnimatedImageViewTarget
import com.majeur.psclient.util.html.Html
import com.majeur.psclient.util.minusFirst
import com.majeur.psclient.widget.BattleLayout
import java.util.concurrent.ExecutionException
import kotlin.math.roundToInt

class GlideHelper(context: Context) {

    enum class SpriteType(private val path: String, private val ext: String) {

        D3ANIMATED("ani", "gif"), // Gen 6+ 3D animated
        D2ANIMATED("gen5ani", "gif"), // Gen 5 2D animated
        D2("gen5", "png"), // Gen 5 2D non animated
        DEX("dex", "png"),
        TRAINER("trainers", "png"); // Dex

        fun uri(spriteId: String, shiny: Boolean, back: Boolean): Uri = Uri.Builder().run {
            scheme("https")
            authority("play.pokemonshowdown.com")
            appendPath("sprites")
            var dir = path
            if (back) dir += "-back"
            if (shiny) dir += "-shiny"
            appendPath(dir)
            appendPath("$spriteId.$ext")
            build()
        }
    }

    companion object {
        private const val MAGIC_SCALE = 0.0027777777777778f
    }

    private val glide = Glide.with(context)

    fun loadBattleSprite(pokemon: BattlingPokemon, imageView: ImageView) {
        val spriteId = pokemon.transformSpecies ?: pokemon.spriteId
        loadSprite(spriteId, pokemon.trainer, pokemon.shiny, true,
                SpriteType.D3ANIMATED, SpriteType.D2ANIMATED, SpriteType.D2)
            .into(object : AnimatedImageViewTarget(imageView) {
            override fun onInitInAnimation(viewPropertyAnimator: ViewPropertyAnimator) {
                viewPropertyAnimator
                        .setDuration(250)
                        .setInterpolator(DecelerateInterpolator())
                        .scaleX(0f)
                        .scaleY(0f)
                        .alpha(0f)
            }

            override fun onInitOutAnimation(viewPropertyAnimator: ViewPropertyAnimator) {
                viewPropertyAnimator
                        .setDuration(250)
                        .setInterpolator(AccelerateInterpolator())
                        .scaleX(1f)
                        .scaleY(1f)
                        .alpha(1f)
            }

            override fun setResource(resource: Drawable) {
                val fieldWidth = (imageView.parent as BattleLayout).width
                var scale = fieldWidth * MAGIC_SCALE
                if (!pokemon.foe) scale *= 1.5f
                imageView.layoutParams.apply {
                    width = (resource.intrinsicWidth * scale).roundToInt()
                    height = (resource.intrinsicHeight * scale).roundToInt()
                }
                imageView.setImageDrawable(resource) // Will request a layout
            }
        })
    }

    fun loadPreviewSprite(player: Player, pokemon: BasePokemon, imageView: ImageView) {
        loadSprite(pokemon.spriteId, player == Player.TRAINER, false, true,
                SpriteType.D3ANIMATED, SpriteType.D2ANIMATED, SpriteType.D2)
                .into(object : AnimatedImageViewTarget(imageView) {
                    override fun onInitInAnimation(viewPropertyAnimator: ViewPropertyAnimator) = Unit
                    override fun onInitOutAnimation(viewPropertyAnimator: ViewPropertyAnimator) = Unit

                    override fun setResource(resource: Drawable) {
                        val fieldWidth = (imageView.parent as BattleLayout?)?.width ?: 0
                        val scale = fieldWidth * MAGIC_SCALE
                        imageView.layoutParams.apply {
                            width = (resource.intrinsicWidth * scale).roundToInt()
                            height = (resource.intrinsicHeight * scale).roundToInt()
                        }
                        imageView.setImageDrawable(resource) // Will request a layout
                    }
                })
    }

    fun loadDexSprite(pokemon: BasePokemon, shiny: Boolean, imageView: ImageView) {
        loadSprite(pokemon.spriteId, false, shiny, true, SpriteType.DEX, SpriteType.D2)
                .into(imageView)
    }

    fun loadAvatar(avatar: String, imageView: ImageView) {
        loadSprite(avatar, false, false, false, SpriteType.TRAINER)
                .into(imageView)
    }

    // Load sprite trying with each sprite type if previous has failed
    private fun loadSprite(spriteId: String, back: Boolean, shiny: Boolean,
                           overrideSize: Boolean, vararg spriteTypes: SpriteType): RequestBuilder<Drawable> {
        val options = RequestOptions().apply {
            if (overrideSize) override(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)
            if (spriteTypes.size == 1) error(R.drawable.missingno) // No more sprite types, default fallback
        }
        return glide.load(spriteTypes.first().uri(spriteId, shiny, back)).apply {
            apply(options)
            if (spriteTypes.size > 1) // There is more sprite types, add an error fallback
                error(loadSprite(spriteId, back, shiny, overrideSize, *spriteTypes.minusFirst()))
        }
    }

    fun getHtmlImageGetter(iconLoader: AssetLoader, maxWidth: Int): Html.ImageGetter {
        val mw = maxWidth - Utils.dpToPx(2f)
        return Html.ImageGetter { source, reqw, reqh ->
            try {
                var d: Drawable? = null
                if (source.startsWith("content://com.majeur.psclient/dex-icon/")) {
                    val species = source.substring(source.lastIndexOf('/') + 1, source.length)
                    val icon = iconLoader.dexIconNonSuspend(species)
                    if (icon != null) d = BitmapDrawable(icon)
                } else {
                    d = glide.asDrawable().load(source).submit().get()
                }
                if (d == null) return@ImageGetter null
                val r = d.intrinsicWidth / d.intrinsicHeight.toFloat()
                var w: Int
                var h: Int
                if (reqw != 0 && reqh == 0) {
                    w = reqw
                    h = (w / r).toInt()
                } else if (reqw == 0 && reqh != 0) {
                    h = reqh
                    w = (h * r).toInt()
                } else {
                    w = reqw
                    h = reqh
                }
                val mr = w / mw.toFloat()
                if (mr > 1) {
                    w = mw
                    h /= mr.toInt()
                }
                d.setBounds(0, 0, w, h)
                return@ImageGetter d
            } catch (e: ExecutionException) {
                e.printStackTrace()
                return@ImageGetter null
            } catch (e: InterruptedException) {
                e.printStackTrace()
                return@ImageGetter null
            }
        }
    }
}